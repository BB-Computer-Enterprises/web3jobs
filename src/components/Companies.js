import { useState, useEffect } from "react";
import { Link } from "react-router-dom";
import { getAllCompaniesInAlphabetic } from "../lib/db";
import {
    COMPANIES_URL,
    COMPANY_DESCRIPTION,
    COMPANY_ID,
    COMPANY_NAME
} from "../lib/constants";
import { makeFriendlyUrl } from "../util/sanitize";

const Companies = () => {
    const [companies, setCompanies] = useState([]);
    const [errorText, setError] = useState("");
    const [isLoading, setIsLoading] = useState(true);

    useEffect(() => {
        fetchCompanies().catch(console.error);
    }, []);

    const fetchCompanies = async () => {
        let { data: companies, error } = await getAllCompaniesInAlphabetic();
        if (error) setError(error);
        else {
            setCompanies(companies)
            setIsLoading(false);
        };
    };

    // function that will destructure the company object
    // it pulls out XXX to be used in the URL
    const generateLinkURL = company => {
        const { [COMPANY_NAME]: cName } = company;
        // return `${JOBS_URL}/${removeWhiteSpace(title)}-${removeWhiteSpace(cName)}/${id}`
        return `${COMPANIES_URL}/${makeFriendlyUrl(cName)}`
    }

    return (
        <div>
            <div className={"w-screen fixed flex flex-col min-h-screen bg-gray-50"}>
                <header
                    className={
                        "flex justify-between items-center px-4 h-16 bg-gray-900"
                    }
                >
                    <span
                        className={
                            "text-2xl sm:text-4xl text-white border-b font-sans"
                        }
                    >
                        Web 3.0 Companies
                    </span>
                </header>
                <div
                    className={"flex flex-col flex-grow p-4"}
                    style={{ height: "calc(100vh - 11.5rem)" }}
                >
                    <div
                        className={`p-2 border flex-grow grid gap-2 ${companies.length ? "auto-rows-min" : ""
                            } grid-cols-1 h-2/3 overflow-y-scroll first:mt-8`}
                    >
                        {companies.length ? (
                            companies.map(company => (
                                <div>
                                    <Link to={{ pathname: generateLinkURL(company), state: { company } }}>
                                        <h1 key={company[COMPANY_ID]}>Company Name: {company[COMPANY_NAME]}</h1>
                                    </Link>

                                    <p>Description: {company[COMPANY_DESCRIPTION]}</p>
                                </div>
                            ))
                        ) : (
                            <span
                                className={
                                    "h-full flex justify-center items-center"
                                }
                            >
                                {isLoading ? 'Loading...' : 'No companies! ðŸ‘€'}
                            </span>
                        )}
                    </div>
                    {!!errorText && (
                        <div
                            className={
                                "border max-w-sm self-center px-4 py-2 mt-4 text-center text-sm bg-red-100 border-red-300 text-red-400"
                            }
                        >
                            {errorText}
                        </div>
                    )}
                </div>
            </div>
        </div>
    );
}

export default Companies;