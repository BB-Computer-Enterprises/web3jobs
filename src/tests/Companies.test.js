import { render, screen } from '@testing-library/react';
import Companies from '../components/Companies.js';

test('Lists the companies if they are available', () => {
    render(<Companies></Companies>);

    const items = screen.findByLabelText('Company Name')
    expect(items)
})
