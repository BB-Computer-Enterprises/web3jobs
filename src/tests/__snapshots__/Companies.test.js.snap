// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests Making sure it doesnt crash 1`] = `
<div>
  <div
    className="w-screen fixed flex flex-col min-h-screen bg-gray-50"
  >
    <header
      className="flex justify-between items-center px-4 h-16 bg-gray-900"
    >
      <span
        className="text-2xl sm:text-4xl text-white border-b font-sans"
      >
        Companies Screen
      </span>
    </header>
    <div
      className="flex flex-col flex-grow p-4"
      style={
        Object {
          "height": "calc(100vh - 11.5rem)",
        }
      }
    >
      <div
        className="p-2 border flex-grow grid gap-2  grid-cols-1 h-2/3 overflow-y-scroll first:mt-8"
      >
        <span
          className="h-full flex justify-center items-center"
        >
          You do have any companies yet!
        </span>
      </div>
    </div>
  </div>
</div>
`;
